[#4 #8] flashcard routes

lilareliga opened the pull request:
Imho gotowe do merge'a

lilareliga committed:
[#4 #8] flashcard routes

CaptainObjective commented on server/src/routes/flashcard.ts:
```diff
@@ -0,0 +1,56 @@
+    const {error} = validateFlashcard(req.body);
+    if(error) return res.status(400).send(error.details[0].message);
+
+    const collection = await FlashcardCollection.findById(req.body.params.id);
```
chyba powinno być req.params.id

Nebrakada commented on server/src/routes/flashcard.ts:
```diff
@@ -0,0 +1,56 @@
+
+router.delete('/:id', async (req: Request, res: Response) => {
+    const flashcard = await Flashcard.findByIdAndRemove(req.params.id);
+    if(!flashcard) return res.status(404).send('The flashcard with the given ID was not found.');
```
przy if zawsze dawajmy { }, kiedyś może dorzucimy coś jeszcze log / debug. nie podkresla tego? nie ma na to reguly w eslint/linterze?

lilareliga replied:
eslint działa jakoś dziwnie, bo miałam import modeli, które nie były zmerdżowane, i niektóre rzeczy mi wypisywał, ale autoformatowanie nie działało.

CaptainObjective replied:
Ja bym się nie upierał przy klamrach w ifach. Poradnik do JS mówi że tak można https://github.com/airbnb/javascript#blocks
Mając backend w node/express taka linia będzie pojawiać się bardzo często we wszystkich endpointach. Decyzje zostawiam wam do przegłosowania.

8bcb replied:
ja nawet uważam że bez klamr jest bardziej elegancko

lilareliga committed:
Merge branch 'develop' into 4-add-flashcard

CaptainObjective commented on server/src/routes/flashcard.ts:
```diff
@@ -0,0 +1,56 @@
+    if(!collection) return res.status(404).send('There is no flashcard collection with a given ID');
+
+    let flashcard = new Flashcard({
+        FlashcardCollection: {//tu pytanie, czy nie potrzebujemy referencji w modelu flashcard
```
Z tego co widzę to narazie FlashcardCollection trzyma FlashCard[] nie bardzo widzę po co nam dodatkowa relacja w drugą stronę wiec powiedziałbym że całe FlashcardCollection tutaj nie potrzebne

CaptainObjective commented on server/src/routes/flashcard.ts:
```diff
@@ -0,0 +1,56 @@
+
+    Flashcard.update({_id: req.params.id}, {$set: updateOps})
+        .exec()
+        .then(result => { 
```
jak już mamy funkcje async to bądźmy konsekwentni i piszmy await :D

CaptainObjective commented on server/src/routes/flashcard.ts:
```diff
@@ -0,0 +1,56 @@
+        extraInfo: req.body.extraInfo
+    }) 
+
+    flashcard = await flashcard.save();
```
Nie bardzo rozumiem po co robimy ponowne przypisanie po zrobieniu save, nie wystarczy po prostu?:  
```suggestion
    await flashcard.save();
```

CaptainObjective commented on server/src/routes/flashcard.ts:
```diff
@@ -0,0 +1,56 @@
+    const flashcard = await Flashcard.findById(req.params.id);
+    if(!flashcard) return res.status(404).send('The flashcard with the given ID was not found.');
+
+    const { error } = await validateFlashcard(req.body); //to dobrze zadziała przy patch? nie udało mi się wygooglować odpowiedzi
```
Nie odpowiem bo nie wiem jak wygląda req.body ;-P

CaptainObjective commented on server/src/routes/flashcard.ts:
```diff
@@ -0,0 +1,56 @@
+    const { error } = await validateFlashcard(req.body); //to dobrze zadziała przy patch? nie udało mi się wygooglować odpowiedzi
+    if(error) return res.status(400).send(error.details[0].message);
+
+    const updateOps = {};
+    for (const ops of req.body) {
+        updateOps[ops.propName] = ops.value
+    }
```
Na ten fragment chciałbym bliżej spojrzeć jak już będą modele i baza danych podpięta. Wygląda jakbyśmy się wystawiali na [prototype pollution](https://portswigger.net/daily-swig/prototype-pollution-the-dangerous-and-underrated-vulnerability-impacting-javascript-applications)

CaptainObjective reviewed (changes requested):
Do merguj proszę developa. Trzeba też spojrzeć na tego twojego lintera bo coś tu ewidentnie nie działa jak należy

lilareliga committed:
[#4 #8] add flashcard routes

- corrections after code review

lilareliga committed:
[#4 #8] add-flashcard

- configuration stuff

CaptainObjective committed:
lila scripts

lilareliga committed:
[#4 #8] add flashcard routes

- fix all routes
- improve error handling

CaptainObjective commented on server/package.json:
```diff
@@ -1,55 +1,58 @@
+        "eslint-plugin-prettier": "^3.3.1",
+        "helmet": "^4.4.1",
+        "jest": "^26.6.3",
+        "nodemon": "^2.0.7",
```
odinstaluj to proszę

lilareliga replied:
@CaptainObjective  bez helmeta i jesta wywalały się testy na githubie:( I też jest problem z przejściem testów na githubie, ale to już wyższa szkoła jazdy. Lokalnie wszystko działa, coś trzeba zmienić tutaj...? Może chodzić o to, że dla testów łączy się z testową bazą na mongodb

CaptainObjective replied:
Chodzi o nodemona

CaptainObjective commented on server/src/routes/flashcard.ts:
```diff
@@ -3,49 +3,55 @@ import { Flashcard, validateFlashcard } from '../models/Flashcard';
+        });
+        const { error } = validateFlashcard(req.body);
+        if (error) return res.status(400).send(error.details[0].message);
+        await collection['flashcards'].push(flashcard);
```
to chyba nie jest asynchroniczne, więc await jest zbędny

CaptainObjective reviewed (commented):
Wygląda że jest git, dopisz proszę jakieś testy

lilareliga committed:
[#4 #8] add flashcard routes

- add tests
- add configuration settings for tests
- update routes in line with test results

lilareliga committed:
[#4 #8] add flashcard routes

- restore jest in package.json

lilareliga committed:
[#4 #8] add flashcard routes

- restore helmet in package.json

lilareliga committed:
[#4 #8] add flashard changes

- restore eslint-plugin-prettier

lilareliga committed:
[#4 #8] add flashcard routes

- add mongoDB in githhub actions

lilareliga committed:
Merge branch 'develop' into 4-add-flashcard

lilareliga committed:
- eslint-related fix to get the tests passed

lilareliga committed:
[#4 #8] add flashcard routes

- uninstall nodemon
